system/controller.go

```diff
@@ -41,3 +44,46 @@ func (controller *Controller) Parse(t *template.Template, name string, data inte
 	}
 	return doc.String()
 }
+
+// CheckPasswordResetToken checks that the input token string is valid,
+// recognized by the DB, and not expired. Flash messages are added to the
+// session for any token failure, and the return indicates if all checks have
+// passed. The UserToken and PasswordReset objects are also returned.
+func (controller *Controller) CheckPasswordResetToken(tokenStr string, c web.C) (models.UserToken, *models.PasswordReset, bool) {
+	session := controller.GetSession(c)
+	dbMap := controller.GetDbMap(c)
+
+	var token models.UserToken
+
+	// Check that the token is set.
+	if tokenStr == "" {
+		session.AddFlash("No password update token present.", "passwordupdateError")
+		return token, nil, false
+	}
+
+	// Check that the token is valid.
+	var err error
+	token, err = models.UserTokenFromStr(tokenStr)
+	if err != nil {
+		session.AddFlash("Email verification token not valid.",
+			"passwordupdateError")
+		return token, nil, false
+	}
+
+	// Check that the token is recognized.
+	passwordReset, err := helpers.PasswordResetTokenExists(dbMap, token)
+	if err != nil {
+		session.AddFlash("Password update token not valid", "passwordupdateError")
+		return token, passwordReset, false
```

Comment:
Should return nil instead of passwordReset, although for non-nil errors it should be nil anyway